File: ArguUser/ArguUserFrontEnd/src/services/commentService.js
/**
 * 댓글 서비스 (Comment Service)
 * 
 * 댓글 관련 API 호출을 담당하는 서비스입니다.
 * 
 * 주요 기능:
 * - 논쟁별 댓글 목록 조회 (페이징)
 * - 댓글 작성
 * - 댓글 삭제
 */

import api from './api'

export const commentService = {
  /**
   * 논쟁별 댓글 목록 조회
   * 
   * 특정 논쟁의 댓글 목록을 페이징하여 가져옵니다.
   * 
   * @param {number} arguId - 논쟁 ID
   * @param {number} page - 페이지 번호 (0부터 시작)
   * @param {number} size - 페이지당 항목 수
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   * @returns {Object} response.data - Page<CommentResponse> (페이징된 댓글 목록)
   */
  async getCommentsByArgu(arguId, page = 0, size = 20) {
    const response = await api.get(`/comments/argu/${arguId}`, {
      params: { page, size },
    })
    return response.data
  },

  /**
   * 댓글 작성
   * 
   * 새로운 댓글을 작성합니다. 인증이 필요합니다.
   * 
   * @param {Object} commentData - 댓글 작성 데이터
   * @param {number} commentData.arguId - 논쟁 ID
   * @param {string} commentData.content - 댓글 내용
   * @param {number} [commentData.parentId] - 부모 댓글 ID (대댓글인 경우)
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   * @returns {Object} response.data - CommentResponse (작성된 댓글 정보)
   */
  async createComment(commentData) {
    const response = await api.post('/comments', commentData)
    return response.data
  },

  /**
   * 댓글 삭제
   * 
   * 댓글을 삭제합니다. 작성자만 삭제 가능합니다.
   * 
   * @param {number} id - 삭제할 댓글 ID
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   */
  async deleteComment(id) {
    const response = await api.delete(`/comments/${id}`)
    return response.data
  },
}

File: ArguUser/ArguUserBackEnd/src/main/java/com/argu/service/CommentService.java
package com.argu.service;

import com.argu.dto.request.CreateCommentRequest;
import com.argu.dto.response.CommentResponse;
import com.argu.entity.Argu;
import com.argu.entity.Comment;
import com.argu.entity.User;
import com.argu.exception.BadRequestException;
import com.argu.exception.ResourceNotFoundException;
import com.argu.repository.ArguRepository;
import com.argu.repository.CommentRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CommentService {
    private final CommentRepository commentRepository;
    private final ArguRepository arguRepository;

    @Transactional
    public CommentResponse createComment(CreateCommentRequest request, Long userId) {
        Argu argu = arguRepository.findById(request.getArguId())
                .orElseThrow(() -> new ResourceNotFoundException("논쟁을 찾을 수 없습니다"));

        User user = new User();
        user.setId(userId);

        Comment parent = null;
        if (request.getParentId() != null) {
            parent = commentRepository.findById(request.getParentId())
                    .orElseThrow(() -> new ResourceNotFoundException("부모 댓글을 찾을 수 없습니다"));

            if (!parent.getArgu().getId().equals(argu.getId())) {
                throw new BadRequestException("부모 댓글이 해당 논쟁에 속하지 않습니다");
            }
        }

        Comment comment = Comment.builder()
                .user(user)
                .argu(argu)
                .parent(parent)
                .content(request.getContent())
                .isHidden(false)
                .build();

        comment = commentRepository.save(comment);
        return CommentResponse.from(comment);
    }

    public Page<CommentResponse> getCommentsByArgu(Long arguId, Pageable pageable) {
        Argu argu = arguRepository.findById(arguId)
                .orElseThrow(() -> new ResourceNotFoundException("논쟁을 찾을 수 없습니다"));

        Page<Comment> comments = commentRepository.findByArguAndIsHiddenFalseAndParentIsNull(argu, pageable);

        return comments.map(comment -> {
            CommentResponse response = CommentResponse.from(comment);
            List<Comment> replies = commentRepository.findByParent(comment);
            response.setReplies(replies.stream()
                    .map(CommentResponse::from)
                    .collect(Collectors.toList()));
            return response;
        });
    }

    @Transactional
    public void deleteComment(Long commentId, Long userId) {
        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new ResourceNotFoundException("댓글을 찾을 수 없습니다"));

        if (!comment.getUser().getId().equals(userId)) {
            throw new BadRequestException("댓글을 삭제할 권한이 없습니다");
        }

        commentRepository.delete(comment);
    }
}


